csharp/src/Ice/CollocatedRequestHandler.cs

```diff
@@ -20,143 +22,95 @@ internal class CollocatedRequestHandler : IRequestHandler
-                        _asyncRequests.Add(requestId, outAsync);
-                    }
+                if (observer != null)
+                {
+                    childObserver = observer.GetCollocatedObserver(_adapter, requestId, outgoingRequestFrame.Size);
+                    childObserver?.Attach();
+                }
 
-                    _sendAsyncRequests.Add(outAsync, requestId);
+                if (_adapter.Communicator.TraceLevels.Protocol >= 1)
+                {
+                    ProtocolTrace.TraceCollocatedFrame(_adapter.Communicator, (byte)Ice1Definitions.FrameType.Request,
+                        requestId, outgoingRequestFrame);
                 }
             }
-            catch (Exception)
-            {
-                _adapter.DecDirectCount();
-                throw;
-            }
 
-            outAsync.AttachCollocatedObserver(_adapter, requestId, outAsync.RequestFrame.Size);
-            if (_adapter.TaskScheduler != null || !synchronous || outAsync.IsOneway || _reference.InvocationTimeout > 0)
+            Task<IncomingResponseFrame?> task;
+            if (_adapter.TaskScheduler != null || !synchronous || oneway || _reference.InvocationTimeout > 0)
             {
                 // Don't invoke from the user thread if async or invocation timeout is set. We also don't dispatch
                 // oneway from the user thread to match the non-collocated behavior where the oneway synchronous
                 // request returns as soon as it's sent over the transport.
-                Task.Factory.StartNew(
-                    () =>
-                    {
-                        if (SentAsync(outAsync))
-                        {
-                            ValueTask vt = InvokeAllAsync(outAsync.RequestFrame, requestId);
-                            // TODO: do something with the value task
-                        }
-                    }, default, TaskCreationOptions.None, _adapter.TaskScheduler ?? TaskScheduler.Default);
+                task = Task.Factory.StartNew(() => InvokeAllAsync(outgoingRequestFrame, requestId), default,
+                    TaskCreationOptions.None, _adapter.TaskScheduler ?? TaskScheduler.Default).Unwrap();
+
+                if (oneway)
+                {
+                    childObserver?.Detach();
+                    return default;
+                }
             }
             else // Optimization: directly call invokeAll
             {
-                if (SentAsync(outAsync))
-                {
-                    Debug.Assert(outAsync.RequestFrame != null);
-                    ValueTask vt = InvokeAllAsync(outAsync.RequestFrame, requestId);
-                    // TODO: do something with the value task
-                }
+                task = InvokeAllAsync(outgoingRequestFrame, requestId);
             }
-        }
+            return new ValueTask<Task<IncomingResponseFrame>?>(WaitForResponseAsync(task, childObserver));
```

Comment:
SendRequestAsync is an asynchronous method that returns a `Task<IncomingResponseFrame>` if the request returns a response (two-way invocation). 

The caller (the proxy InvokeAsync implementation) can then wait on this task to wait for the response. I think it is better to keep the responsibility of waiting for the response in the proxy code rather than in the request handler SendRequestAsync method which is only responsible today for sending the request.

The proxy code needs to know when the request is sent to call the Progress callback and also for retries to check if it's safe to retry for at-most-once semantics (the setting of the sent bool in the InvokeAsync implementation). If SendRequestAsync was both responsible for sending and waiting for the response, we'd somehow would have to return this flag with the connection's exception if there's a failure...